mg.c

```diff
@@ -945,6 +931,14 @@ Perl_magic_get(pTHX_ SV *sv, MAGIC *mg)
             if (dwErr) {
                 PerlProc_GetOSError(sv, dwErr);
                 fixup_errno_string(sv);
+
+#     ifdef USE_LOCALE
+                if (   IN_LOCALE
+                    && get_win32_message_utf8ness(SvPV_nomg_const_nolen(sv)))
+                {
+                    SvUTF8_on(sv);
+                }
+#     endif
```

Comment:
Isn't fixup_errno_string() already doing something similar to this?

That said the selected locale doesn't seem to change the messages anyway:
```
#!perl
use Devel::Peek;
use POSIX qw(LC_ALL setlocale);
use locale;
my $loc = setlocale(LC_ALL, "Japanese.utf8");
print $loc, "\n";
unlink "Doesn't exist";
my $x = $^E;
Dump($x);
my $y = $!;
Dump($y);
```

result (I have the Japanese language pack installed):
```
C:\Users\tony\dev\perl\git\perl\win32>..\perl ..\..\errortest.pl
Japanese_Japan.utf8
SV = PVMG(0x17d4ba24d18) at 0x17d4ba0c860
  REFCNT = 1
  FLAGS = (NOK,POK,IsCOW,pNOK,pPOK,UTF8)
  IV = 0
  NV = 2
  PV = 0x17d4d53fb78 "The system cannot find the file specified"\0 [UTF8 "The system cannot find the file specified"]
  CUR = 41
  LEN = 43
  COW_REFCNT = 1
SV = PVMG(0x17d4ba24d48) at 0x17d4d4a2580
  REFCNT = 1
  FLAGS = (NOK,POK,IsCOW,pNOK,pPOK)
  IV = 0
  NV = 2
  PV = 0x17d4d4ea798 "No such file or directory"\0
  CUR = 25
  LEN = 27
  COW_REFCNT = 1
```

Reply:
IIUC, you're suggesting moving the check into the fixup function. I think it's slightly better the current way.  Moving it there would mean that would have an #ifdef for Windows, instead of here, and that would be the only case where that function would care about UTF8ness.  But that function is called from another place where UTF8ness has already been determined before the call.  And that would mean an inconsistent use of that function.  So I think it's better to keep that function not worrying about UTF-8 in any circumstance, and leave the needed special case here.

It has been my experience that most language packs do not come with the error strings in anything but English.  I just checked on the only box currently available to me located in a country where the native language isn't English, and the message was in English even so.  These commits were developed a couple of years ago, and I did find boxes where some locales did have translations for these.  Perhaps they were @Tux 's boxes which are currently offline

Reply:
The messages from $^E seem to depend on the system language, when I changed the session language to Japanese in the control panel (which required a log out / in), I instead got:
```
C:\Users\tony\dev\perl\git\perl\win32>..\perl -w ..\..\errortest.pl
Japanese_Japan.utf8
SV = PVMG(0x20e5d2c6f78) at 0x20e5d2acff0
  REFCNT = 1
  FLAGS = (NOK,POK,IsCOW,pNOK,pPOK,UTF8)
  IV = 0
  NV = 2
  PV = 0x20e5ed05038 "??????????????????"\0 [UTF8 "??????????????????"]
  CUR = 18
  LEN = 20
  COW_REFCNT = 1
SV = PVMG(0x20e5d2c6fa8) at 0x20e5ecae468
  REFCNT = 1
  FLAGS = (NOK,POK,IsCOW,pNOK,pPOK)
  IV = 0
  NV = 2
  PV = 0x20e5ed06db8 "No such file or directory"\0
  CUR = 25
  LEN = 27
  COW_REFCNT = 1
```

This wasn't locale dependent. commenting out the setlocale() made no difference.

As to the fixup function, it was more that the extra code added here and the fixup function have a very strong overlap, the new code comes close to setting UTF-8 in all the circumstances the fixup function does., so perhaps the call to the fixup could be eliminated.

An alternative would be to make PerlProc_GetOSError() (aka win32_str_os_error()) always return the UTF-8 version of the message, and flag its return value that way (since it works with an SV rather than a char *).

Reply:
@leont, do you have an opinion on the 2nd paragraph of Tony's comment